### ISSUE TITLE
Detach tensor from graph after grad evaluation in denoiser

### ISSUE THREAD

**Issue by fedor-goncharov** [deepinv/deepinv#548]
In GSPnP class the `potential_grad` method turns on `requires_grad` but does not sets its off after. 

```python
def potential_grad(self, x, sigma, *args, **kwargs):
        r"""
        Calculate :math:`\nabla g` the gradient of the regularizer :math:`g` at input :math:`x`.

        :param torch.Tensor x: Input image
        :param float sigma: Denoiser level :math:`\sigma` (std)
        """
        with torch.enable_grad():
            x = x.float()
            x = x.requires_grad_()
            N = self.student_grad(x, sigma)
            JN = torch.autograd.grad(
                N, x, grad_outputs=x - N, create_graph=True, only_inputs=True
            )[0]
        Dg = x - N - JN
        return self.alpha * Dg
```

Maybe, in general, it is safer to call `x.detach()` after? (it may eat a lot of memory if unnoticed I think)

**Comment by mh-nguyen712** [comment]
Hi @fedor-goncharov and thanks for spotting this out. Calling `x.detach()` will break the computational graph and one can not differentiate (a second time) through the forward, for example when training an unfolded model or even training GSPnP denoiser. 
Most of the algorithm for reconstruction are implemented with torch.no_grad() already.

**Comment by fedor-goncharov** [comment]
@mh-nguyen712 I see, maybe it is worth to add a flag "detach" in the method? In unrolling you set it to True but for 
other methods you can set it to False. I think yet it is a "hidden feature" that returned tensor is requiring a grad when you may not expect it. 

I think fast prototyping is really  a cool feature of this library, maybe even more important than having many algorithms implemented inside. My comment was if I decide to 'play' with denoised tensor I might accidentally eat all my gpu memory ... :)

**Comment by mh-nguyen712** [comment]
Could be a good option!
